WcaOnRails/db/migrate/20170517213035_update_list_of_countries.rb

```diff
@@ -0,0 +1,54 @@
+# frozen_string_literal: true
+
+class UpdateListOfCountries < ActiveRecord::Migration[5.0]
+  def up
+    Country.delete_all
+    Country::ALL_STATES.each do |state|
+      Country.create(state)
+    end
+    # These substitions have been found by running the migration and checking
+    # users with iso2 not matching anything in the 'Country' table.
+    {
+      # Virgin Islands -> United Kingdom
+      "VG" => "GB",
+      # Aruba -> Netherlands
+      "AW" => "NL",
+      # French Guinea -> France
+      "GF" => "FR",
+      # Pitcairn Islands -> United Kingdom
+      "PN" => "GB",
+      # Puerto Rico -> USA
+      "PR" => "US",
+      # Isle of Man -> United Kingdom
+      "IM" => "GB",
+      # French Polynesia -> France
+      "PF" => "FR",
+    }.each do |old_iso2, new_iso2|
+      User.where(country_iso2: old_iso2).update(country_iso2: new_iso2)
+    end
+    {
+      "Aruba" => "Netherlands",
+      "Puerto Rico" => "USA",
+    }.each do |old_id, new_id|
+      Person.where(countryId: old_id).update(countryId: new_id)
+      Result.where(countryId: old_id).update(countryId: new_id)
+    end
+
+    problems = []
+    problems << User.where.not(country_iso2: Country.uncached_real.map(&:iso2))
+    problems << Person.where.not(countryId: Country.uncached_real.map(&:id))
+    problems.flatten!
+    problematic_results = Result.where.not(countryId: Country.uncached_real.map(&:id))
+    unless problems.empty? && problematic_results.empty?
```

Comment:
This may make the migration harder to write, but is is possible to do these checks upfront? My concern with doing them after the database work is that it means the migration will fail, and leave the database in a possibly weird state. It would be nice if this migration could be written in a way such that it either all happens, or all does not happen.

I always wondered why cleaning up after failed migrations is such a mess, it seems like ActiveRecord should just wrap every migration in a transaction and then the migrations can either happen, or not happen. Then I remembered that MySQL does not support transactions around creating or deleting tables, which is oftentimes what a migration does. It turns out the PostgreSQL *does* support what are called "database definition language (DDL) transactions", so this would behave better on PostgreSQL (see this marvelous SO answer explaining all that https://stackoverflow.com/a/29943069).

Since this migration is *not* altering our database schema in any way, I think it would be safe to wrap this whole `def up` inside of a `ActiveRecord::Base.transaction`, which might be the simplest way of ensuring that this migration either entirely happens or does not happen at all.